### PR TITLE
BUG: indexing with boolean-like Index, #11119

### LINKED ISSUE #11119: Slicing multiple DataFrame columns doesn't work with boolean column names
Setup: 

```
import pandas as pd
df = pd.DataFrame({True: [1,2,5,9],
                   False: [6,1,13,8]})
```

`df[True]` and `df[False]` work fine; but `df[[True,False]]` throws an exception:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-a28f55cd0379> in <module>()
----> 1 df[[True,False]]

/Users/username/pandasenv/lib/python2.7/site-packages/pandas/core/frame.pyc in __getitem__(self, key)
   1772         if isinstance(key, (Series, np.ndarray, Index, list)):
   1773             # either boolean or fancy integer index
-> 1774             return self._getitem_array(key)
   1775         elif isinstance(key, DataFrame):
   1776             return self._getitem_frame(key)

/Users/username/pandasenv/lib/python2.7/site-packages/pandas/core/frame.pyc in _getitem_array(self, key)
   1809             elif len(key) != len(self.index):
   1810                 raise ValueError('Item wrong length %d instead of %d.' %
-> 1811                                  (len(key), len(self.index)))
   1812             # _check_bool_indexer will throw exception if Series key cannot
   1813             # be reindexed to match DataFrame rows

ValueError: Item wrong length 2 instead of 4.
```

On the other hand, this works fine:

```
df = pd.DataFrame({'True': [1,2,5,9],
                   'False': [6,1,13,8]})
df[['True','False']]
```

Probably relevant to: https://github.com/pydata/pandas/issues/9595

### PR DESCRIPTION
**Pull request by preddy5**
closes #11119

### CODE REVIEW

**Comment by pradyu1993** [comment]
@jorisvandenbossche

**Comment by maximilianr** [comment]
Does this only work where you supply two items for a DataFrame, because it's checking for `ndim=len(key)`? Do you want to be checking the length of the axis?

```python
In [11]: df2=pd.concat([df, pd.Series([1,3,12,9],name=True)],axis=1)

In [12]: df2
Out[12]: 
   False  True   True 
0      6      3      1
1      1      9      3
2     13      8     12
3      8      2      9


In [13]: df2[[False]]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-13-b7f220d1c4ee> in <module>()
----> 1 df2[[False]]

/Users/maximilian/Dropbox/workspace/pandas/pandas/core/frame.py in __getitem__(self, key)
   1906         if isinstance(key, (Series, np.ndarray, Index, list)):
   1907             # either boolean or fancy integer index
-> 1908             return self._getitem_array(key)
   1909         elif isinstance(key, DataFrame):
   1910             return self._getitem_frame(key)

/Users/maximilian/Dropbox/workspace/pandas/pandas/core/frame.py in _getitem_array(self, key)
   1947                 else:
   1948                     raise ValueError('Item wrong length %d instead of %d.' %
-> 1949                                  (len(key), len(self.index)))
   1950             # check_bool_indexer will throw exception if Series key cannot
   1951             # be reindexed to match DataFrame rows

ValueError: Item wrong length 1 instead of 4.

In [17]: df2[[False, True, False]]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-88caa367ce94> in <module>()
----> 1 df2[[False, True, False]]

/Users/maximilian/Dropbox/workspace/pandas/pandas/core/frame.py in __getitem__(self, key)
   1906         if isinstance(key, (Series, np.ndarray, Index, list)):
   1907             # either boolean or fancy integer index
-> 1908             return self._getitem_array(key)
   1909         elif isinstance(key, DataFrame):
   1910             return self._getitem_frame(key)

/Users/maximilian/Dropbox/workspace/pandas/pandas/core/frame.py in _getitem_array(self, key)
   1947                 else:
   1948                     raise ValueError('Item wrong length %d instead of %d.' %
-> 1949                                  (len(key), len(self.index)))
   1950             # check_bool_indexer will throw exception if Series key cannot
   1951             # be reindexed to match DataFrame rows

ValueError: Item wrong length 3 instead of 4.
```

While if the column names are strings:
```python
In [33]: df2_str[['False','True']]
Out[33]: 
   False  True  True
0      6     3     1
1      1     9     3
2     13     8    12
3      8     2     9
```

**Comment by jreback** [comment]
What actually is needed is to infer this:

```
In [22]: pd.lib.infer_dtype(Index([True,False]))
Out[22]: 'boolean'
```

so right now we don't have a separate boolean ``Index`` type, its just ``object`` dtype.

So this particular case only is relevant when you: 

- have a boolean indexer
- have a boolean ``Index`` on the same axis (so it must be ``object`` dtype), then you infer

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1941,7 +1941,11 @@ def _getitem_array(self, key):
                 warnings.warn("Boolean Series key will be reindexed to match "
                               "DataFrame index.", UserWarning)
             elif len(key) != len(self.index):
-                raise ValueError('Item wrong length %d instead of %d.' %
+                if lib.infer_dtype(self._info_axis)=="boolean":
```
Comment (pradyu1993) [inline pandas/core/frame.py]: @jreback am I doing it right ?

Comment (jreback) [inline pandas/core/frame.py]: I think you can move this check to ``_convert_to_indexer``, and raise if their is a problem (e.g. subsume the ``ValueError`` there). makes this code a lot simpler. right now too much special casing.

**Comment by pradyu1993** [comment]
@jreback can you help me understand the error, I am unable to reproduce the error on my notebook and the error is only occurring for python2.7

**Comment by jreback** [comment]
its unrelated, a spurious failed. I restarted. In any event I will look at this prob next week. You are jumping thru lots of hoops here.

**Comment by pradyu1993** [comment]
@jreback Could you review the PR. Thanks

**Comment by shoyer** [comment]
I'm not sure if this change is actually a good idea. Suppose we have a 2x2 DataFrame with columns and rows `[True, False]`:
```
In [3]: df = pd.DataFrame([[1, 2], [3, 4]], columns=[True, False], index=[True, False])

In [4]: df
Out[4]:
       True   False
True       1      2
False      3      4
```

What should `df[[False, True]]` return? Standard indexing with `[]` is a mess of fallback rules, so usually it's best to first think about what `.loc` and `.iloc` can handle.

Here, note that both `.loc` and `.iloc` [claim to support](http://pandas.pydata.org/pandas-docs/stable/indexing.html#different-choices-for-indexing) boolean arrays as indexers. Given that indexes can include booleans, this is clearly an ambiguous case in our current rules. The current behavior may be a bug, but all the indexers seem to currently use booleans arrays for subsetting rows, not looking up index values:
```
In [5]: df[[True, False]]
Out[5]:
      True   False
True      1      2

In [9]: df.loc[[True, False]]
Out[9]:
      True   False
True      1      2
```

So my suggestion is that we should first consider deprecating using booleans that are not index values with `.loc`. Then, once this is entirely unambiguous, we can consider appropriate fallback rules for normal indexing.

**Comment by jreback** [comment]
deprecating boolean indexers for ``.loc`` is a non-starter as a) this would break all back-compat, b) this would make ``.loc`` completely non-intuitive and unusable with a boolean array, which is a core feature.

Supporting label indexing is currently a buggy edge case. I'll review soon.

**Comment by shoyer** [comment]
I agree that this would be a backwards incompatibility break, but a pretty smooth deprecation cycle would be possible. I disagree that it is unintuitive for `.loc` to not work with boolean arrays, because `.loc` is explicitly for labeled based indexing. We've simply gotten away with it because using booleans for labels is pretty rare.

**Comment by jreback** [comment]
@shoyer I don't see any good reason to not allow ``.loc`` to accept boolean arrays. This would cause way more issues than it solves. In fact it is *quite* common to do this. Not sure why you are pushing this way. We want *more* consistency and unification, not less.

**Comment by jreback** [comment]
@pradyu1993 rebase this on master pls.

this needs a fair amount more testing. (of other cases of boolean labels).

**Comment by jreback** [comment]
closing. if you want to update according to comments. pls reopen

**Pull request closed without merging**